internal/controllers/extension_controller.go

```diff
@@ -156,10 +249,161 @@ func (r *ExtensionReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		case declcfg.SchemaPackage:
+			apimeta.SetStatusCondition(&ext.Status.Conditions, metav1.Condition{
+				Type:               ocv1alpha1.TypePackageDeprecated,
+				Reason:             ocv1alpha1.ReasonDeprecated,
+				Status:             metav1.ConditionTrue,
+				Message:            deprecation.Message,
+				ObservedGeneration: ext.Generation,
+			})
+		case declcfg.SchemaChannel:
+			if ext.Spec.Source.Package.Channel != deprecation.Reference.Name {
+				continue
+			}
+
+			apimeta.SetStatusCondition(&ext.Status.Conditions, metav1.Condition{
+				Type:               ocv1alpha1.TypeChannelDeprecated,
+				Reason:             ocv1alpha1.ReasonDeprecated,
+				Status:             metav1.ConditionTrue,
+				Message:            deprecation.Message,
+				ObservedGeneration: ext.Generation,
+			})
+		case declcfg.SchemaBundle:
+			apimeta.SetStatusCondition(&ext.Status.Conditions, metav1.Condition{
+				Type:               ocv1alpha1.TypeBundleDeprecated,
+				Reason:             ocv1alpha1.ReasonDeprecated,
+				Status:             metav1.ConditionTrue,
+				Message:            deprecation.Message,
+				ObservedGeneration: ext.Generation,
+			})
+		}
+
+		deprecationMessages = append(deprecationMessages, deprecation.Message)
+	}
+
+	if len(deprecationMessages) > 0 {
+		apimeta.SetStatusCondition(&ext.Status.Conditions, metav1.Condition{
+			Type:               ocv1alpha1.TypeDeprecated,
+			Reason:             ocv1alpha1.ReasonDeprecated,
+			Status:             metav1.ConditionTrue,
+			Message:            strings.Join(deprecationMessages, ";"),
+			ObservedGeneration: ext.Generation,
+		})
+	}
+}
+
+// findStatusCondition finds the conditionType in conditions.
+// TODO: suggest using upstream conditions to Carvel.
+func findStatusCondition(conditions []kappctrlv1alpha1.Condition, conditionType kappctrlv1alpha1.ConditionType) *kappctrlv1alpha1.Condition {
+	for i := range conditions {
+		if conditions[i].Type == conditionType {
+			return &conditions[i]
+		}
+	}
+	return nil
+}
+
+func (r *ExtensionReconciler) ensureApp(ctx context.Context, desiredApp *unstructured.Unstructured) error {
+	existingApp, err := r.existingAppUnstructured(ctx, desiredApp.GetName(), desiredApp.GetNamespace())
+	if client.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	// If the existing BD already has everything that the desired BD has, no need to contact the API server.
+	// Make sure the status of the existingBD from the server is as expected.
+	if equality.Semantic.DeepDerivative(desiredApp, existingApp) {
+		*desiredApp = *existingApp
+		return nil
+	}
+
+	return r.Client.Patch(ctx, desiredApp, client.Apply, client.ForceOwnership, client.FieldOwner("operator-controller"))
+}
+
+func (r *ExtensionReconciler) existingAppUnstructured(ctx context.Context, name, namespace string) (*unstructured.Unstructured, error) {
+	existingApp := &kappctrlv1alpha1.App{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, existingApp)
```

Comment:
I recall that the controller-runtime client has some potentially unexpected default behavior around unstructured objects (iirc, the unstructured client is uncached by default).

Looks like we setup a structured informer for Apps in like 252.  Is it possible to fetch structured and convert to unstructured without breaking the comparison logic? Alternatively, could we setup the watch to use unstructured Apps and configure the client to enable caching of unstructured Apps?

Reply:
Nevermind! I completely missed this being a structured Get converted to unstructured. All good.